src/main/scala/diplomaticobjectmodel/OMLogicalTree.scala

```diff
@@ -0,0 +1,72 @@
+// See LICENSE.SiFive for license details.
+
+package freechips.rocketchip.tile
+
+import freechips.rocketchip.diplomacy.ResourceBindingsMap
+import freechips.rocketchip.diplomaticobjectmodel.model.OMComponent
+import scala.collection.mutable.ArrayBuffer
+
+
+trait LogicalTree {
+  def getOMComponents(resourceBindingsMap: ResourceBindingsMap, children: Seq[OMComponent]): Seq[OMComponent]
+}
+
+/** LogicalTreeEdges hold LogicalTree Nodes which will be used to construct the logical tree.
+  * The LogicalTreeEdge is used to construct a parent child relationship between different modules.
+  *
+  * @param parent The parent LogicalTree node
+  * @param child The child LogicalTree node
+  */
+case class LogicalTreeEdge(
+  parent: LogicalTree,
+  child: LogicalTree
+)
+
+object LogicalModuleTree {
+  val edges = ArrayBuffer[LogicalTreeEdge]()
+
+  def add(parent: LogicalTree, child: LogicalTree): Unit = {
+     edges += LogicalTreeEdge(parent, child)
+  }
+
+  /**
+    * The getTreeMap function constructs a tree by converting the list of LogicalTreeEdges into a Map which
+    * contains each of the parents and their children
+    *
+    * @return
+    */
+  def getTreeMap(): Map[LogicalTree, List[LogicalTree]] = {
+    edges.groupBy(_.parent).map{ case (k, v) => (k, v.map(_.child).toList)}
+  }
+
+  def findRoot(): LogicalTree = {
+    val values = getTreeMap().values.flatten.toSet
+    val keys = getTreeMap().keys.toSet
+
+    val roots = values.diff(keys)
+
+    assert(roots.size == 1, "Logical Tree contains more than one root.")
+    roots.head
+  }
+}
+
+case class Tree[A](parent: A, children: List[Tree[A]])
+
+object OMLogicalTreeTree {
```

Comment:
Why is this its own object instead of just some more methods on `LogicalModuleTree`?

Reply:
1. First a list of   val edges = ArrayBuffer[LogicalTreeEdge]() is constructed.

2. Then def getTreeMap(): Map[LogicalTree, List[LogicalTree]] converts the list into a map which represents the tree.

3. Then the tree map is converted into a OMLogicalTree by the  def makeTree(): Tree[LogicalTree].

The division between the tree map and OMLogical tree was done to separate the concerns.

Reply:
So there are three or four data structures here, some of which have their own object whose only function is to convert it to another. I don't understand why there are so many.

1. Couldn't you just serialize out the edge-map structure itself? Why do we need to turn it into a actual immutable nested tree data structure? If downstream tools are going to ask graphical questions about it, they are probably just going to turn it back into an edge map. Plus, it looks like only the final `Seq[OMComponents]` gets serialized anyways, so why even do all these conversions?
2. What value does `Tree[LogicalTree]` add over `Map[LogicalTree, List[LogicalTree]]`? If the latter is too long a type name, you can define a local type member `type LogicalTree = Map[LogicalTreeNode, Seq[LogicalTreeNode]]`
3. Why doesn't `OMLogicalTree.makeTree()` just take the `resourceBindings` as an argument? As is, both `OMLogicalTree.makeTree` and `OMTree.tree` both have to recurse through the whole tree, once each.

Overall I think `Map[LogicalTreeNode, List[LogicalTreeNode]]` is probably the fundamental data being captured here, and I think if you used a mutable Map and re-wrote the `add` function you could build it as you go instead of computing it later. Ultimately, the only thing being serialized is the `Seq[OMComponent]`, so it seems sufficient to do:

```
object OMLogicalTree {
  private val tree  = mutable.Map[LogicalTreeNode, Seq[LogicalTreeNode]].empty
  def add(edge: (LogicalTreeNode, LogicalTreeNode)): Unit = {
    tree += ((edge._1, tree(edge._1) + edge._2) // ... or something like this 
  }
  def roots: Seq[LogicalTreeNode] = { ... }
  def checkIsTree: Boolean = roots.size == 1
  def bind(resourceBindingsMap: ResourceBindingsMap): Seq[OMComponent] = {
    def recurse(node: LogicalTreeNode): Seq[OMComponent] = 
       node.getOMComponents(resourceBindingsMap, tree(node).flatMap(recurse))
    recurse(root)
  }
}
```

Reply:
Certainly your solution works given the current of requirements. I am leaving room for expansion later to create graphs, such as interrupt, clock, power,  between the tree nodes.

Reply:
If they are Graphs, why are they called Trees?